### PR TITLE
test: Add adaptive quantum phase estimation examples

### LINKED ISSUE #299: Create example program: Adaptive Quantum Phase Estimation
e.g. https://arxiv.org/abs/2206.12950

### PR DESCRIPTION
**Pull request by mark-koch**
Closes #299

### CODE REVIEW

**Comment by croyzor** [comment]
I thought that the scope of these "example" tasks was that they'd serve more as documentation, and go in the top level "examples" folder

File: tests/integration/test_examples.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+from guppylang.module import GuppyModule
+from guppylang.prelude.builtins import result
+from guppylang.prelude.quantum import qubit
+
+import guppylang.prelude.quantum as quantum
+from guppylang.prelude.quantum import h, cx, x, measure, rz, reset
+
+
+def test_random_walk_phase_estimation(validate):
+    # Implementation of the adaptive random walk phase estimation algorithm from
```
Comment (croyzor) [inline tests/integration/test_examples.py:13]: We can use docstrings for this since #334

File: tests/integration/test_examples.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        while i < num_iters:
+            aux = h(qubit())
+            t = 1 / sigma
+            aux = rz(h(aux), (sigma - mu) * t)
+            aux, q = controlled_oracle(aux, tgt, t)
+            if measure(h(aux)):
+                mu += sigma * 0.6065  # = sigma/sqrt(e)
+            else:
+                mu -= sigma * 0.6065  # = sigma/sqrt(e)
+            sigma *= 0.7951  # = sqrt((e-1)/e)
```
Comment (croyzor) [inline tests/integration/test_examples.py:52]: My point about constants applies doubly since they're defined elsewhere and inlined here separately!

File: tests/integration/test_examples.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+from guppylang.decorator import guppy
+from guppylang.module import GuppyModule
+from guppylang.prelude.builtins import result
+from guppylang.prelude.quantum import qubit
+
+import guppylang.prelude.quantum as quantum
+from guppylang.prelude.quantum import h, cx, x, measure, rz, reset
+
+
+def test_random_walk_phase_estimation(validate):
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply tests/integration/test_examples.py:12]: Yeah, good catch. Turns out the program has few issues when actually trying to compile it :D

Will push some fixes

File: tests/integration/test_examples.py (hunk 1 above)
Comment (mark-koch) [reply tests/integration/test_examples.py:13]: The Guppy functions below already have docstrings (thanks btw!). Not sure if people usually add docstrings to pytest functions?

File: guppylang/prelude/builtins.py [hunk 4]
```diff
@@ -424,7 +424,7 @@ def __nat__(self: float) -> nat: ...
     def __ne__(self: float, other: float) -> bool: ...
 
     @guppy.hugr_op(builtins, float_op("fneg"), CoercingChecker())
-    def __neg__(self: float, other: float) -> float: ...
+    def __neg__(self: float) -> float: ...
```
Comment (mark-koch) [inline guppylang/prelude/builtins.py:427]: Drive-by fix

File: tests/integration/test_examples.py (hunk 1 above)
Comment (croyzor) [reply tests/integration/test_examples.py]: Oops, i wasn't paying close enough attention to what kind of function i was looking at!

It just seemed odd to open a python function with a descriptive comment vs a docstring even if it's not library code 🤷 of course, i'm no expert on python conventions

**Review by croyzor — APPROVED** [review]
